### DISCUSSION TITLE
Why are old files automatically stored with Git LFS without migration?

### DISCUSSION THREAD

**Discussion by ArtoriasDelAbismo** [git-lfs/git-lfs#5919] [Q&A]
I recently started using Git LFS and followed the documentation, which states that only new files matching the tracked patterns will be stored with Git LFS. To include existing files, the documentation mentions using git lfs migrate.

Here’s what I did:

Installed Git LFS and tracked these pattern:

**git lfs track "*.jpg"**

followed by:

**git add .gitattributes**

Ran **git add -A** and noticed older files committed and pushed long ago in the staging area. Committed and pushed to the main repository. When I checked GitHub, older files already in the repository were marked as 'Stored with Git LFS', even though I didn’t run git lfs migrate.

My understanding was that older files wouldn’t be affected unless migrated. So my question is:

- How could older files already be stored with Git LFS if migration wasn’t performed?

- Is the documentation incorrect or am I misunderstanding how Git LFS handles old files?

Thanks for any clarification!

**Comment by chrisd8088** [comment]
Hey, thanks for the question!

I think what's going on here is that you used the `-A` option to `git add`, without any paths, which means all files in your working tree will be updated in the Git index, and so your new `.gitattributes` patterns will apply and existing files will be run through Git LFS's `clean` filter, creating Git LFS objects.  From the Git [documentation](https://git-scm.com/docs/git-add#Documentation/git-add.txt--A):

> If no `<pathspec>` is given when `-A` option is used, all files in the entire working tree are updated

However, your Git commit history will not have been changed, so if you examine a previous commit, those files won't have been converted to Git LFS objects in those commits.  That's what `git lfs migrate import` would do, though; it would rewrite that history.

In other words, there's a difference between "older files" meaning "files which are already in your working tree", which I think is the sense in which you're using the term, and meaning "files which exist in prior commits".

Here's an example to clarify the situation.  Note how, before the `git lfs migrate import`, the commit prior to the current `HEAD` just has `foo.bin` as a regular file, not a Git LFS object, and there's no `.gitattributes` file in that prior commit.  After running the command, the commit history has changed, and the `HEAD` commit now changes no files at all; the `.gitattributes` file is introduced in the prior commit, and `foo.bin` is a Git LFS object from that commit onwards.

```
$ git init test
$ cd test
$ echo foo >foo.bin
$ git add foo.bin
$ git commit -m foo

$ git lfs track '*.bin'
$ git add -A
$ git commit -m attrs

$ git lfs ls-files
b5bb9d8014 * foo.bin
$ git show HEAD:foo.bin
version https://git-lfs.github.com/spec/v1
oid sha256:b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
size 4

$ git lfs ls-files HEAD^
[no output]
$ git show HEAD^:foo.bin
foo

$ git show --name-only --oneline HEAD
f42180a (HEAD -> main) attrs
.gitattributes
foo.bin
$ git show --name-only --oneline HEAD^
5dbe6d8 foo
foo.bin

$ git lfs migrate import --include '*.bin'

$ git lfs ls-files HEAD^
b5bb9d8014 * foo.bin
$ git show HEAD^:foo.bin
version https://git-lfs.github.com/spec/v1
oid sha256:b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
size 4

$ git show HEAD^:.gitattributes
*.bin filter=lfs diff=lfs merge=lfs -text

$ git show --name-only --oneline HEAD
2536740 (HEAD -> main) attrs
$ git show --name-only --oneline HEAD^
e7a49f4 foo
.gitattributes
foo.bin
```

**Reply by ArtoriasDelAbismo** [reply]
Thank you for the quick and detailed response! It clarified a lot about how Git LFS processes files and the distinction between the working tree and prior commits.

I ran git lfs ls-files, and the output was as follows:

e77d4c7c2f * public/open-graph.png
7ef8c5e0b4 * public/project.png
a7089d7d1f * src/assets/images/capsule_616x353.jpg
882423f303 * src/assets/images/compass-1.jpg
03e2507434 * src/assets/images/compass.jpg
8d1bbdb9a1 * src/assets/images/hallucinations.jpeg
6685cd2a31 * src/assets/images/hellisus.jpg
8a739368f3 * src/assets/images/netlify-cover.jpg
2dad6f9ea7 * src/assets/images/self-development.jpeg
85e076959d * src/assets/images/ssh.png

Of these, the only file I added after running git lfs track and adding .gitattributes was src/assets/images/hellisus.jpg.

Based on your explanation, I expected that the output of **git show HEAD^:src/assets/images/compass.jpg**, for instance, would return the contents of a regular file (since it existed prior to adding Git LFS tracking). However, the actual output is:

version https://git-lfs.github.com/spec/v1
oid sha256:a7089d7d1f7d117fd138b992a8fc1c6cc6c1ed064b6ce1f66c41a5334dae59df
size 66950

This behavior seems to contradict the idea that older files (in prior commits) are unaffected unless git lfs migrate import is used. Could this be related to how Git LFS handles updates to the index after git add -A, or might there be an additional step or nuance I’m missing?

Once again, I appreciate your insights, your explanation already helped me understand a lot better how Git LFS operates!

**Reply by chrisd8088** [reply]
Here's what I can say with certainty: `git lfs migrate` is the _only_ Git LFS command which rewrites your Git history.

But there are lots of Git commands which do!

Perhaps you used `git commit --amend`, or in some other way applied the changes staged in the Git index by `git add` (with or without `-A`) to earlier commits?

We'd need to see the exact series of commands you used to give a precise answer, though.

**Comment by MV10** [comment]
The OP is right. 

I _exactly_ followed the instructions on the LFS web page (install, track, add) ... it will modify ALL of the pre-existing tracked files.
